cpp/src/arrow/builder.cc

```diff
@@ -1222,9 +1222,14 @@ Status BinaryBuilder::Init(int64_t elements) {
 Status BinaryBuilder::Resize(int64_t capacity) {
   DCHECK_LT(capacity, std::numeric_limits<int32_t>::max());
   // one more then requested for offsets
-  RETURN_NOT_OK(offsets_builder_.Resize((capacity + 1) * sizeof(int64_t)));
+  RETURN_NOT_OK(offsets_builder_.Resize((capacity + 1) * sizeof(int32_t)));
   return ArrayBuilder::Resize(capacity);
 }
+  
+Status BinaryBuilder::ReserveData(int64_t capacity) {
+  DCHECK_LT(capacity, std::numeric_limits<int32_t>::max());
+  return value_data_builder_.Resize(capacity * sizeof(int64_t));
```

Comment:
Why do we multiply here with `int64_t`? I would expect that `ReserveData(x)` will lead to `value_data_capacity() = x`.

Reply:
I think this should be `Resize(capacity + value_data_builder_.length())`. The `sizeof(int64_t)` looks like a copy-paste error from implementation of `BinaryBuilder::Resize` above (and that should be `sizeof(int32_t)` there, so we should fix that)

We should check that `extra_capacity + length` does not exceed INT32_MAX but probably return `Status::Invalid` since overflowing a `BinaryBuilder` is likely to happen somewhat more regularly

I note also that `BufferBuilder` and `TypedBufferBuilder<T>` don't have a `shrink_to_fit` option in their `Resize` method, that would be good to add to avoid unnecessary reallocations

Reply:
Opened https://issues.apache.org/jira/browse/ARROW-2004

Reply:
Thank you two for the comments.

@xhochy you are right. I will make the change accordingly.

@wesm I have modified BinaryBuilder::Resize to use sizeof(int32_t) in previous commit. Will change to Resize(extra_capacity + curr_length)